### ISSUE TITLE
Layering / Venue

### ISSUE THREAD

**Issue by litherum** [tc39-transfer/proposal-intl-segmenter-v2#8]
From the perspective of Apple's WebKit team:

# Background

[Intl.Segmenter v1](https://tc39.es/proposal-intl-segmenter/) was promoted to Stage 3 recently. It includes facilities to iterate over grapheme cluster boundaries, word boundaries, and sentence boundaries. We see reasonable use-cases for all of these kinds of segments: find-in-page probably wants to search word-by-word, and triple-click probably wants to select text one-sentence-at-a-time. @Constellation has [implemented support](https://webkit.org/blog/11648/new-webkit-features-in-safari-14-1/) in WebKit. So far so good.

# This proposal

This ECMA402 proposal is about expanding the existing segmented support to include line breaking. We view this as a philosophically different proposal, despite being implementable using the same infrastructure, for a few reasons:

1. This API reports where it is linguistically legal to break a line, not where the line breaks actually occur. The other boundary types report where the boundaries actually occur.
2. Every use case for this API involves passing the results of this API to text width measurement APIs, and thus any callers will need to make sure to have correct behavior for font files / metrics, font fallback, bidi reordering, shaping, etc.
3. Line breaking is a key piece in authors trying to perform paragraph-level text layout themselves in script, without using any of the browser’s existing functionality. This kind of reimplementation is usually less correct than the browser’s built-in layout facilities, and usually less performant.

Now, we don’t want to pass judgment on whether or not JavaScript authors _should_ be performing paragraph-level layout themselves in script - we won’t say anything about that. But we do need to recognize that it is a fact that authors do try to perform line breaking in JavaScript. We’ve heard anecdotal stories of authors simply marking every space character as a line breaking opportunity. Chrome’s [usage counters show](https://chromestatus.com/metrics/feature/timeline/popularity/3177) that authors use their non-standard v8BreakIterator. Google Charts breaks long chart labels using v8BreakIterator in Chrome, but simply ellipsizes the same labels in other browsers. Doing the job correctly, for certain scripts like Burmese, Khmer, Lao, and Thai requires [megabytes of dictionaries](https://github.com/unicode-org/icu/tree/main/icu4c/source/data/brkitr/dictionaries).

# Position and Rationale

Because of the above points, we feel the proposed [Intl.Segmenter v2 API](https://github.com/tc39-transfer/proposal-intl-segmenter-v2/blob/main/README.md) is a poor fit in the context of the existing web platform, for these reasons:

1. **Insufficiency**: All the stated use cases involve using the result of the API to pick a returned breaking opportunity that causes text to fill an available space. The JavaScript standard library does not include any facilities for fonts or text measurement.
2. **Layering**: Fundamentally, the problem of line breaking is a text layout problem, which is handled in the web platform at the canvas, DOM, and CSS layers, rather than the JavaScript layer.
3. **Performance / Correctness**: Browsers are in a unique position among the community of different language runtimes in that they already have first-rate text layout engines inside them already. Browsers’ text engines have a close interrelationship between font fallback, bidi processing, line breaking, and shaping, which cannot be matched using this API proposal.

Our position is: if line breaking functionality is added to the web platform, it should be added in a higher-level API, which can properly solve the larger problem of paragraph-level text layout, including all of its facets such as font fallback, glyph metrics, shaping, bidi, etc, in a way that can use the browser’s existing text layout engine. (Again - we are not saying that line breaking _should_ be added to the web platform; but only that if it is added, it should be added accordingly.)

# Potential Paths Forward

One satisfactory example would be if the CSS Houdini group pursued a potential Custom Text Layout API as an effort similar to its ongoing [Custom Layout API](https://drafts.css-houdini.org/css-layout-api/). Another even more preferable example would be to augment the existing canvas text APIs within the [Canvas Formatted Text](https://github.com/WICG/canvas-formatted-text/blob/main/README.md) effort. (Indeed, the Canvas Formatted Text effort [describes in its readme](https://github.com/WICG/canvas-formatted-text/blob/main/README.md#goals) how it seems to be trying to solve exactly this line breaking problem!)

**Comment by FrankYFTang** [comment]
> but simply ellipsizes the same labels in other browsers.
Actually, what happen is the chart "simply ellipsizes the same labels in ALL browsers" for the text which is too long to break, INCLUDING on chrome. For example, if you have a very long string as lable "long abcdefghijklmnopqrstuvwxyzasdfadfafdsafafdfasfs", it will also show the "ellipsizes" on chrome because chrome cannot break that long *word*. The showing of ellipsizes is on all browser for the part the text cannot be break. And the reason it is only showing up on browser other than chrome is because there are no API on other browser to break the text.

**Comment by FrankYFTang** [comment]
> 1. **Insufficiency**: ...The JavaScript standard library does not include any facilities for fonts or text measurement.

The **Insufficiency** part in some rendering context is standarized by w3c and implemented by Safari and all other browser already:

https://developer.apple.com/documentation/webkitjs/canvasrenderingcontext2d/1631690-measuretext
https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/measureText
https://developer.mozilla.org/en-US/docs/Web/API/TextMetrics

**Comment by FrankYFTang** [comment]
> 2\. **Layering**: Fundamentally, the problem of line breaking is a text layout problem, which is handled in the web platform at the canvas, DOM, and CSS layers, rather than the JavaScript layer.

I agree w/ you there is a need for provide a higher level solution, but in all platforms we study, it does not exclude the need to have a low level API. Here is several exames:

1. MacOS API
   * Line break in low level API - in [Core Foundation](https://developer.apple.com/documentation/corefoundation?language=objc) "Access low-level functions, primitive data types, and various collection types that are bridged seamlessly with the Foundation framework." 
     * [CFStringTokenizerCreate](https://developer.apple.com/documentation/corefoundation/1542047-cfstringtokenizercreate?language=objc) w/ [kCFStringTokenizerUnitLineBreak](https://developer.apple.com/documentation/corefoundation/1588024-tokenization_modifiers/kcfstringtokenizerunitlinebreak?language=objc)

     * [UCFindTextBreak](https://developer.apple.com/documentation/coreservices/1390429-ucfindtextbreak/) w/ [kUCTextBreakLineMask](https://developer.apple.com/documentation/coreservices/1390433-anonymous/kuctextbreaklinemask?language=objc) 
   * Text layout in higher level API - in [Core Text](https://developer.apple.com/library/archive/documentation/StringsTextFonts/Conceptual/CoreText_Programming/Introduction/Introduction.html?language=objc#//apple_ref/doc/uid/TP40005533)
     * [CTFrameDraw](https://developer.apple.com/library/archive/documentation/StringsTextFonts/Conceptual/CoreText_Programming/LayoutOperations/LayoutOperations.html#//apple_ref/doc/uid/TP40005533-CH12-SW11)
   
MacOS clearly has BOTH levels of API co-exist, a low level API needed inside [Core Foundation](https://developer.apple.com/documentation/corefoundation?language=objc) which has no information of font or way to measure text, and also in the higer level API [Core Text](https://developer.apple.com/library/archive/documentation/StringsTextFonts/Conceptual/CoreText_Programming/Introduction/Introduction.html ) which has font and a way to measure text. MacOS APIs show a very common pattern of release about this in the industry - have BOTH a lowel one which provide the layout opportunity and a higher level one for common usage.

2. Java API:
   * Line break in low level API - in [ inside J2SE java.text package](https://docs.oracle.com/javase/7/docs/api/java/text/package-summary.html#package_description) in [java.text.BreakIterator](https://docs.oracle.com/javase/8/docs/api/java/text/BreakIterator.html) 
   * Text layout in higher level API
     * in [[old] AWT]() "... API to develop Graphical User Interface (GUI) or windows-based applications in Java."- use [LineBreakMeasurer](https://docs.oracle.com/javase/tutorial/2d/text/drawmulstring.html) or [TextLayout](https://docs.oracle.com/javase/8/docs/api/java/awt/font/TextLayout.html)
     * In [Swing components](https://docs.oracle.com/javase/tutorial/uiswing/components/index.html) : "GUI widget toolkit for Java" - use  [JTextComponent](https://docs.oracle.com/javase/8/docs/api/javax/swing/text/JTextComponent.html)

3. Android API:
   * Line break in low level API - in [android.icu.text  package](https://developer.android.com/reference/android/icu/text/package-summary) Use [android.icu.text.BreakIterator](https://developer.android.com/reference/android/icu/text/BreakIterator) 
   * Text layout in higher level API - in [android.widget package](https://developer.android.com/guide/topics/appwidgets/overview) - use [android.widget.TextView](https://developer.android.com/reference/android/widget/TextView)

As you can see, BOTH levels of API co-exist  in a platform is a very standard practice in the industry for many years.

**Comment by FrankYFTang** [comment]
There are no reason for the desire of a higher level API solution exclude the need of a lower level API solution. And the higher level API solution may not solve all the use case for the ECMAScript charter to solve (notice ECMAScript is a language standard, a lower level and a bigger scope standard than the "web platform") We are not proposing this proposal to be added to "web platform" (which is ECMAScript used in web) here, we are proposing this proposal to be added to ECMAScript, which an implementatino could exsit in EITHER for the web OR  without html layout engine built in) A higher level API won't be easily implement in that context (non web  ECMAScript usage, such as jsPDF running on a edge device which has access to font info but no HTML layout but only has PDF layout)

**Comment by FrankYFTang** [comment]
I accidentally hit the close button 11 mins ago so I click the reopen button right after.

**Comment by FrankYFTang** [comment]
@yjbanov

**Comment by gibson042** [comment]
I remain convinced that there are relevant concerns here for non-browser environments which should be addressed at the level of ECMA-402. It does not preclude anything in Houdini, but nor does anything in Houdini preclude pursuing this proposal.

**Comment by litherum** [comment]
> ONLY a (possible multi years) higher level api solution

This is not what we're advocating for. Under the assumption that this problem should be solved (which we are not saying either way about), we're advocating for a higher-level solution first, then followed if necessary by a lower-level solution if the use cases of the lower-level solution cannot fit into the higher-level solution.

**Comment by litherum** [comment]
@FrankYFTang you've mentioned that some proposals might take multiple years. Can you elaborate about the urgency of this API request?

There hasn't _ever_ been a standard way to break lines on the web without using the DOM, so I'm confused why this request seems urgent. For a platform as long-lived as the web, getting a feature right the first time is much more important than a solution that can ship in the shortest amount of time. (I'm coming from the perspective that line breaking is a fundamentally different problem than segmenting by grapheme cluster, words, or sentences, for the reasons given in the above post.)

**Comment by litherum** [comment]
> I remain convinced that there are relevant concerns here for non-browser environments

What are the concerns? Non-browser environments don't browse the web (presumably?), so their compat requirements are quite different. Node.js, for example, can use a native module to provide line breaking, and wouldn't need this addition to the JS standard library.

**Comment by ljharb** [comment]
@litherum most code used in node is deployed both in node and in browsers, and thus most node code benefits greatly from virtually everything browsers use. Additionally, serverside rendering means *every single HTML generating use case* that exists in the browser, applies equally to node, full stop.

**Comment by FrankYFTang** [comment]
> Can you elaborate about the urgency of this API request?

**Comment by FrankYFTang** [comment]
> Can you elaborate about the urgency of this API request?

Because we have multiple internal and external customers request the low level one.

**Comment by FrankYFTang** [comment]
> There hasn't _ever_ been a standard way to break lines on the web without using the DOM

[UAX 14](https://unicode.org/reports/tr14/) is "a standard way to break lines on the web without using the DOM" for more than 20 years. Applying UAX14 insdie a &lg;canvas&gt; involve no DOM inside &lg;canvas&gt; and only need to use standard html &lg;canvas&gt; api .

**Comment by othermaciej** [comment]
The existence of requests from unnamed customers shouldn't determine the venue or the shape of the API. It's not right to rush a standard that is architecturally wrong (and quite difficult, maybe impossible to use fully correct). If the customers want to show up and participate in the discussion and explain why their use cases could only be solved by this (arguably) poorly layered API and not by a more appropriate higher level API like Canvas Formatted Text, then that would be worth considering. It's hard to debate with people or entities who are not present.

**Comment by gibson042** [comment]
How would someone use Canvas Formatted Text without Canvas?

**Comment by litherum** [comment]
> How would someone use Canvas Formatted Text without Canvas?

I think that's up to the designers of the API.

**Comment by ljharb** [comment]
@litherum ok, but if the designers of the API don't provide a way to use it without Canvas, then the existence of Canvas Formatted Text must not preclude a more generally accessible API.

**Comment by tabatkins** [comment]
The Canvas Formatted Text and [Houdini Font Metrics API](https://drafts.css-houdini.org/font-metrics-api/) were written with the assumption of being part of the web platform, and thus make some assumptions about the availability of things like the DOM. That said, nothing about the proposals *requires* this to be the case. In particular, the Font Metrics API has a `.measureText()` method that takes a string and some styling information; it's currently located on `Document` (but that's for convenience, not necessity) and takes a CSSOM object (same), both of which could be changed to make the API amenable to use in Node or similar.

In other words, the fully integrated APIs that Myles and Maciej are asking for already mostly exist in spec form, and need just small tweaks to ensure they're usable for generic JS. We don't need to pretend we're starting from scratch.

**Comment by yjbanov** [comment]
## Hi!

I'm here to represent a named customer: Flutter. We're more of a meta-customer, as Flutter is a _framework_, not an _app_. Our customers are app developers though. The proposed API has the potential to reduce the payloads of Flutter apps targeting web by 0.5 MB (gzipped). The size of CanvasKit currently bundled with Flutter weighs 2.8 MB (gzipped), which is too large for many use-cases. We know how to remove some of that. Switching to brotli, taking advantage of `ImageDecoder`, and lazy-loading rarely used feature gets us down to 2.23 MB. Within these 2.23 MB roughly 1 MB is the text system. Within that, 0.5 MB is ICU that serves text segmentation. Ultimately, we'd like most of that 1 MB eliminated. We think we can get to something close to 1.2 MB _total_, and that's not including size optimizing the core functionality.

How did we end up here? Because we render to WebGL, and just like the proposed API, it's not a complete solution to the larger problem of rendering graphics. You still need help from external (and large, and complex) libraries to do anything significant in WebGL (and not shoot yourself in the foot). And that's OK! One step at a time. The proposed API seems like a logical next step.

## Problem statement

I think one thing that deserves attention in this discussion is the question of how diverse text layout is. My view is that it's not one monolithic problem, and therefore is not solvable by one monolithic solution. "Paragraph-level text layout" is mentioned, but what is a "paragraph". Consider these examples: [1](https://www.bwillcreative.com/wp-content/uploads/2020/07/how-to-wrap-text-in-photoshop-hero-image.jpg), [2](https://i.ytimg.com/vi/2LrFz2QZHXs/maxresdefault.jpg), [3](https://i.ytimg.com/vi/K2dh-6NT3s0/maxresdefault.jpg), [4](https://helpx.adobe.com/content/dam/help/en/indesign/using/threading-text/jcr%3acontent/main-pars/image_0/iu_01.png.img.png), [5](https://flutter.github.io/assets-for-api-docs/assets/painting/strut_force_example_2.png).

Therefore I believe the focus should be on _primitives_ that developers need to build a "paragraph".

## Sub-problems

@litherum and I have had a long discussion of the different components of text layout and their interactions in https://github.com/tc39/proposal-intl-segmenter-v2/issues/10. PTAL.

As mentioned above, line breaking is too diverse to cover with one API. Trying to come up with One Universal Paragraph is a string one could pull their entire life. Rendering is as diverse, if not more, due to several rendering technologies available - SVG, HTML, Canvas, WebGL, WebGPU - and of those WebGL and WebGPU offering unlimited customization opportunities.

As of right now, line breaking and rendering are not strong candidates for standardization. I would argue the two components that are worth addressing at the standard API level is **text segmentation** and (potentially) **text shaping**.

**Text segmentation**: this is what @FrankYFTang is proposing here. We do need to make sure the final result can be used to lay out text correctly and in all languages. We do need to make sure this API composes with the rest of the system efficiently. We can't cut corners. However, we do not need another browser API for that. We could take a few customers who express the need for this API and have them build on it and verify the results are good.

**Text shaping**: my initial intuition was that text shaping would be another candidate for standardization, but since @litherum pointed me to https://harfbuzz.github.io/harfbuzzjs/hb.wasm, which compresses to a 91 KB (gzipped) wasm library, I'm not 100% sure any more. We'll need to investigate. It's unclear if that library cuts corners it shouldn't be cutting. **Notably**, there are use-cases where text shaping is exercised without segmentation, such as in widgets that force their text to never line break, and to implement features such as [strut style](https://api.flutter.dev/flutter/painting/StrutStyle-class.html).

Given this breakdown of the problem, a standalone text segmentation API seems appropriate. I do not see layering issues.

## TC39 vs W3C

Flutter doesn't care. We'll take it from whoever offers it. I looked at both https://github.com/tc39/proposal-intl-segmenter-v2/issues/9 and https://github.com/tc39/proposal-intl-segmenter-v2/issues/10 in depth, and I cannot spot a conflict with this being part of TC39. @FrankYFTang's proposal seems to satisfy the provided examples (yes, a little extra coding is required, but again, let's think _primitives_; browser vendors are not the only ones who can write code).

Of all the arguments, the strongest combination I see that would push it towards W3C is if both of the following are true:

* We find that text segmentation is _only_ useful to lay out text and nothing else.
* We find that text shaping also needs standardization.

Then perhaps this API alone would feel incomplete. We may also lose efficiencies from having one set of people address the bigger problem. That's not a technical constraint though.

## A little bureaucracy

A call to invite more customers to the discussion makes sense, the earlier the better. Technically, that only becomes a requirement somewhere on the boundary between Stage 2 and 3 in the [TC39 process](https://tc39.es/process-document/) (see "Indicate that further refinement will require feedback from implementations and users" as the purpose of Stage 3, followed up by "significant in-the-field experience with shipping implementations, such as that provided by two independent VMs" as acceptance criteria of Stage 4). @FrankYFTang's request right now is only to advance to Stage 2.

Having said that, if anyone's in touch with other potential customers, I think we should send them invites.
